apps/vectordb/tests/test_vectorstore.py

```diff
@@ -0,0 +1,143 @@
+    assert output == [Document(page_content="foo", metadata={"page": "0"})]
+
+
+@pytest.mark.django_db()
+def test_vectorstore_with_metadatas_with_scores(experiment):
+    """Test end to end construction and search."""
+    texts = ["foo", "bar", "baz"]
+    metadatas = [{"page": str(i)} for i in range(len(texts))]
+    docsearch = PGVector.from_texts(
+        texts=texts,
+        embedding=DeterministicFakeEmbedding(size=ADA_TOKEN_COUNT),
+        metadatas=metadatas,
+        experiment=experiment,
+        distance_strategy=DistanceStrategy.COSINE,
+    )
+    output = docsearch.similarity_search_with_score("foo", k=1)
+    _remove_meta_fields(output)
+    assert output == [(Document(page_content="foo", metadata={"page": "0"}), 0.0)]
+
+
+@pytest.mark.django_db()
+def test_vectorstore_with_filter_match(experiment):
+    """Test end to end construction and search."""
+    texts = ["foo", "bar", "baz"]
+    metadatas = [{"page": str(i)} for i in range(len(texts))]
+    docsearch = PGVector.from_texts(
+        texts=texts,
+        embedding=DeterministicFakeEmbedding(size=ADA_TOKEN_COUNT),
+        metadatas=metadatas,
+        experiment=experiment,
+    )
+    output = docsearch.similarity_search_with_score("foo", k=1, filter={"page": "0"})
+    _remove_meta_fields(output)
+    assert output == [(Document(page_content="foo", metadata={"page": "0"}), 0.0)]
+
+
+@pytest.mark.django_db()
+def test_vectorstore_with_filter_distant_match(experiment):
+    """Test end to end construction and search."""
+    texts = ["foo", "bar", "baz"]
+    metadatas = [{"page": str(i)} for i in range(len(texts))]
+    docsearch = PGVector.from_texts(
+        texts=texts,
+        embedding=DeterministicFakeEmbedding(size=ADA_TOKEN_COUNT),
+        metadatas=metadatas,
+        experiment=experiment,
+    )
+    output = docsearch.similarity_search_with_score("foo", k=1, filter={"page": "2"})
+    _remove_meta_fields(output)
+    # ordering here is deterministic but random due to fake embeddings
+    assert output == [(Document(page_content="baz", metadata={"page": "2"}), 0.9290842232061864)]
+
+
+@pytest.mark.django_db()
+def test_vectorstore_distant_match_cosine_ordering(experiment):
+    texts = ["foo", "bar", "baz"]
+    metadatas = [{"page": str(i)} for i in range(len(texts))]
+    docsearch = PGVector.from_texts(
+        texts=texts,
+        embedding=DeterministicFakeEmbedding(size=ADA_TOKEN_COUNT),
+        metadatas=metadatas,
+        experiment=experiment,
+        distance_strategy=DistanceStrategy.COSINE,
+    )
+    output = docsearch.similarity_search_with_score("foo", k=3)
+    _remove_meta_fields(output)
+    # ordering here is deterministic but random due to fake embeddings
+    assert output == [
+        (Document(page_content="foo", metadata={"page": "0"}), 0.0),
+        (Document(page_content="baz", metadata={"page": "2"}), 0.9290842232061864),
+        (Document(page_content="bar", metadata={"page": "1"}), 1.0246011368123038),
```

Comment:
Interesting that this is larger than 1. I presume these scores refer to angles then?

Reply:
I was a bit surprised by this too but then I checked and pgvector returns the cosine distance and not the cosine similarity (https://github.com/pgvector/pgvector/blob/d7354a86a8496a4432b9d27e2eaf5658274ed1f4/src/vector.c#L696)

### tl;dr

Similarity is the angle which ranges from -1 to 1
Distance is 1 - similarity which ranges from 0 to 2

See https://medium.com/@milana.shxanukova15/cosine-distance-and-cosine-similarity-a5da0e4d9ded for a good article on the difference.

Reply:
That makes sense, thanks!